core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -3177,13 +3177,15 @@ object Stream {
       * Note that resulting stream won't contain the `head` of this leg.
       */
     def stream: Stream[F, O] =
-      Stream.fromFreeC(Algebra.setScope(scope.id).flatMap { _ =>
-        next
-      })
+      Pull
```

Comment:
Could you describe this change a little? I don't understand the need for looping here and it seems like we should still be consulting `next` at some point.

Reply:
The `next` is consulted by using the `leg.stepLeg` on each loop. This way we can step through to the end of the leg without the need for special algebra that will set the scope in the interpreter. There should not be much of overhead, if any, doing it this way instead of via the `SetScope` algebra.